### PR TITLE
doc/rados: edit bluestore-config-ref.rst (1 of x)

### PR DESCRIPTION
**Pull request by zdover23**
Edit the first part of doc/rados/configuration/bluestore-config-ref.rst.

https://tracker.ceph.com/issues/58485

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 1]
```diff
@@ -1,84 +1,92 @@
...
-devices.
+BlueStore manages either one, two, or in certain cases three storage devices.
 
-In the simplest case, BlueStore consumes a single (primary) storage device.
-The storage device is normally used as a whole, occupying the full device that
-is managed directly by BlueStore. This *primary device* is normally identified
-by a ``block`` symlink in the data directory.
+In the simplest case, BlueStore consumes a single storage device. The device is
+known as the *primary device*. Any device used by BlueStore will be used
+entirely by BlueStore and nothing other than BlueStore. The primary device is
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:12]: I'm a bit uncertain about this rewording, I think it diverges from the intent of the original and in some situations is misleading.  I propose:

>BlueStore manages either one, two, or in certain cases three storage devices.  These _devices_ are in the Linux / Unix sense of something that appears under `/dev` or `/devices`, and may be an entire storage drive, a partition of a storage drive, or a logical volume.  BlueStore does not create or mount a conventional filesystem on devices that it uses, it read and writes to them directly in a "raw" fashion.

>In the simplest case, BlueStore consumes all of a single storage device, which we call the `primary device`.  The primary device is identified by the ``block`` symlink in the metadata directory.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 2]
```diff
@@ -1,84 +1,92 @@
...
+identified by a ``block`` symlink in the data directory.
 
-The data directory is a ``tmpfs`` mount which gets populated (at boot time, or
-when ``ceph-volume`` activates it) with all the common OSD files that hold
-information about the OSD, like: its identifier, which cluster it belongs to,
-and its private keyring.
+The data directory is a ``tmpfs`` mount. When this data directory is booted or
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:15]: I want to call this the *metadata* directory, but if *data* is entrenched, so be it.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 3]
```diff
@@ -1,84 +1,92 @@
...
+activated by ``ceph-volume``, it is populated with all of the common OSD files
+that hold information about the OSD: for example, the OSD's identifier, the
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:17]: populated with metadata files and links that hold

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 4]
```diff
@@ -1,84 +1,92 @@
...
-It is also possible to deploy BlueStore across one or two additional devices:
+In more complicated cases, BlueStore is deployed across one or two additional
+devices:
 
-* A *write-ahead log (WAL) device* (identified as ``block.wal`` in the data directory) can be
-  used for BlueStore's internal journal or write-ahead log. It is only useful
-  to use a WAL device if the device is faster than the primary device (e.g.,
-  when it is on an SSD and the primary device is an HDD).
+* A *write-ahead log (WAL) device* (identified as ``block.wal`` in the data
+  directory) can be used for BlueStore's internal journal or write-ahead log.
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:24]: can be used to separate out BlueStore's internal

What I'm trying to do here is to reinforce the idea that WAL and DB aren't optional things to add to an OSD, but rather things that by default are colocated with the primary data but may be split out.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 5]
```diff
@@ -1,84 +1,92 @@
...
 .. _bluestore-single-type-device-config:
 
 **block (data) only**
 ^^^^^^^^^^^^^^^^^^^^^
-If all devices are the same type, for example all rotational drives, and
-there are no fast devices to use for metadata, it makes sense to specify the
-block device only and to not separate ``block.db`` or ``block.wal``. The
-:ref:`ceph-volume-lvm` command for a single ``/dev/sda`` device looks like:
+If all devices are of the same type (for example, they are all rotational
+drives), and if there are no fast devices available for the storage of
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:77]: We call them HDDs above ....

What I Believe:  they should be placed on a pedastal (so we can look up their platters) and be referred to only as .... spinners.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 6]
```diff
@@ -1,84 +1,92 @@
...
 
    ceph-volume lvm create --bluestore --data /dev/sda
 
-If logical volumes have already been created for each device, (a single LV
-using 100% of the device), then the :ref:`ceph-volume-lvm` call for an LV named
-``ceph-vg/block-lv`` would look like:
+If each device already has a logical volume that has been created for that
+device and that logical volume is using 100% of the device, then the
+:ref:`ceph-volume-lvm` call for an logical volume named ``ceph-vg/block-lv`` is
+as follows:
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:89]: There's been a lot of confusion over this.  I *think* is is about pre-created LVMs, so maybe

>If the devices to be used for a BlueStore OSD are pre-created logical volumes, then the

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 7]
```diff
@@ -88,15 +96,19 @@ using 100% of the device), then the :ref:`ceph-volume-lvm` call for an LV named
 
 **block and block.db**
 ^^^^^^^^^^^^^^^^^^^^^^
-If you have a mix of fast and slow devices (SSD / NVMe and rotational),
-it is recommended to place ``block.db`` on the faster device while ``block``
-(data) lives on the slower (spinning drive).
 
-You must create these volume groups and logical volumes manually as 
-the ``ceph-volume`` tool is currently not able to do so automatically.
+If you have a mix of fast and slow devices (for example, SSD or NVMe and
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:100]: NVMe (mostly) implies SSD.

>for example, SSD or HDD

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 8]
```diff
@@ -88,15 +96,19 @@ using 100% of the device), then the :ref:`ceph-volume-lvm` call for an LV named
...
+rotational drive).
 
-For the below example, let us assume four rotational (``sda``, ``sdb``, ``sdc``, and ``sdd``)
-and one (fast) solid state drive (``sdx``). First create the volume groups:
+You must create these volume groups and these logical volumes manually. as The
+``ceph-volume`` tool is currently unable to do so [create them?] automatically.
+
+The following procedure illustrates the manual creation of volume groups and
+logical volumes.  For this example, we shall assume four rotational drives
+(``sda``, ``sdb``, ``sdc``, and ``sdd``) and one (fast) solid state drive
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:110]: SSD

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 9]
```diff
@@ -114,8 +126,9 @@ Now create the logical volumes for ``block``:
    lvcreate -l 100%FREE -n block-2 ceph-block-2
    lvcreate -l 100%FREE -n block-3 ceph-block-3
 
-We are creating 4 OSDs for the four slow spinning devices, so assuming a 200GB
-SSD in ``/dev/sdx`` we will create 4 logical volumes, each of 50GB:
+Because there are four rotational devices, there will be four OSDs. Supposing
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:129]: If someone ever uses a DVD-RAM drive for an OSD we can and should slap them, so let's call them HDDs consistently.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 10]
```diff
@@ -134,54 +147,57 @@ Finally, create the 4 OSDs with ``ceph-volume``:
    ceph-volume lvm create --bluestore --data ceph-block-2/block-2 --block.db ceph-db-0/db-2
    ceph-volume lvm create --bluestore --data ceph-block-3/block-3 --block.db ceph-db-0/db-3
 
-These operations should end up creating four OSDs, with ``block`` on the slower
-rotational drives with a 50 GB logical volume (DB) for each on the solid state
-drive.
+After this procedure is finished, there should be four OSDs, ''block'' should
+be on the four rotational drives, and each rotational drive should have a 50GB
+logical volume (specifically, a DB device) on the solid state drive.
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:152]: shared SSD

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 11]
```diff
@@ -134,54 +147,57 @@ Finally, create the 4 OSDs with ``ceph-volume``:
...
+
+In older releases, internal level sizes are such that the DB can fully utilize
+only those specific partition / logical volume sizes that correspond to sums of
+L0, L0+L1, L1+L2, and so on--that is, given default settings, sizes of roughly
+3GB, 30GB, 300GB, and so on. Most deployments do not substantially benefit from
+sizing that accommodates L3 and higher, though DB compaction can be facilitated
+by doubling these figures to 6GB, 60GB, and 600GB.
+
+Improvements in Nautilus 14.2.12, Octopus 15.2.6, and subsequent releases allow
+for better utilization of arbitrary DB-device sizes. Moreover, the Pacific
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:177]: maybe arbitrarily-sized DB devices.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 12]
```diff
@@ -233,66 +248,65 @@ The data fraction can be calculated by
...
 .. confval:: bluestore_csum_type
 
-Inline Compression
+inline compression
 ==================
 
-BlueStore supports inline compression using `snappy`, `zlib`, or
-`lz4`. Please note that the `lz4` compression plugin is not
+bluestore supports inline compression using `snappy`, `zlib`, or
+`lz4`. please note that the `lz4` compression plugin is not
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:284]: HUH? News to me, I'm using it right now.  Maybe this was once true.  Curiously `test/compressor/test_compression.cc` appears to not test `zstd`.

>BlueStore offers optional inline compression using `snappy`, `zlib`, `zstd`, or`lz4`.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 13]
```diff
@@ -315,36 +329,36 @@ set with:
...
 
-Internally BlueStore uses multiple types of key-value data,
-stored in RocksDB.  Each data type in BlueStore is assigned a
-unique prefix. Until Pacific all key-value data was stored in
-single RocksDB column family: 'default'.  Since Pacific,
-BlueStore can divide this data into multiple RocksDB column
-families. When keys have similar access frequency, modification
-frequency and lifetime, BlueStore benefits from better caching
-and more precise compaction. This improves performance, and also
+internally bluestore uses multiple types of key-value data,
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:335]: Why lowercase the heading? and the follow sentences?

BlueStore stores multiple types of key-value data in an internal RocksDB database.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 14]
```diff
@@ -315,36 +329,36 @@ set with:
...
+bluestore can divide this data into multiple rocksdb column
+families. when keys have similar access frequency, modification
+frequency and lifetime, bluestore benefits from better caching
+and more precise compaction. this improves performance, and also
 requires less disk space during compaction, since each column
 family is smaller and can compact independent of others.
 
-OSDs deployed in Pacific or later use RocksDB sharding by default.
-If Ceph is upgraded to Pacific from a previous version, sharding is off.
+osds deployed in pacific or later use rocksdb sharding by default.
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:346]: case?

**Review by anthonyeleven — CHANGES_REQUESTED** [review]
A number of comments.  Notably at one point it looks like you inadvertently lowercased the remainder of the file.

**Comment by zdover23** [comment]
> A number of comments. Notably at one point it looks like you inadvertently lowercased the remainder of the file.

I don't know what happened to cause the rest of the file to become lowercase, but I've remedied that in another PR, which I will make sure overwrites the edit to the last half of the file introduced in this PR. Here's the PR that fixes the rest of the file: https://github.com/ceph/ceph/pull/51787

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 1 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:12]: Largely accepted with very minor changes.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:17]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 4 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:24]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 5 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:77]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 6 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:89]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 9 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:129]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 10 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:152]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 11 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:177]: Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 12 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:286]: This was a change introduced on 02 Aug 2017. Given that (1) that is nearly six years ago, and (2) in this domain, you are the boots-on-the-ground non-plus-ultra dealer of in-the-field expertise, we are going to remove the note that says that the ``lz4`` compression plugin is not supported. But we're going to remove the note in https://github.com/ceph/ceph/pull/51787, to keep things tidier and easier for me.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 15]
```diff
@@ -134,54 +149,57 @@ Finally, create the 4 OSDs with ``ceph-volume``:
    ceph-volume lvm create --bluestore --data ceph-block-2/block-2 --block.db ceph-db-0/db-2
    ceph-volume lvm create --bluestore --data ceph-block-3/block-3 --block.db ceph-db-0/db-3
 
-These operations should end up creating four OSDs, with ``block`` on the slower
-rotational drives with a 50 GB logical volume (DB) for each on the solid state
-drive.
+After this procedure is finished, there should be four OSDs, ''block'' should
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:152]: Should those single quotes be backticks?

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 14 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:348]: All of this weird e e cummings/bell hooks-looking stuff is repaired in https://github.com/ceph/ceph/pull/51787.

File: doc/rados/configuration/bluestore-config-ref.rst [hunk 16]
```diff
@@ -195,34 +213,33 @@ used as fallbacks.
...
+``bluestore_cache_size_ssd`` if the primary device is an SSD.
 
-BlueStore and the rest of the Ceph OSD daemon do the best they can
-to work within this memory budget.  Note that on top of the configured
-cache size, there is also memory consumed by the OSD itself, and
-some additional utilization due to memory fragmentation and other
-allocator overhead.
+BlueStore and the rest of the Ceph OSD daemon make every effort to work within
+this memory budget. Note that in addition to the configured cache size, there
+is also memory consumed by the OSD itself. There is alos some additional
```
Comment (anthonyeleven) [inline doc/rados/configuration/bluestore-config-ref.rst:225]: s/alos some//

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 15 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:152]: They sure should.

Accepted.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 16 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:225]: Good God. Good catch.

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 12 above)
Comment (anthonyeleven) [reply doc/rados/configuration/bluestore-config-ref.rst:286]: Sure.  Be sure to add `zstd` there as well.

**Review by anthonyeleven — APPROVED** [review]

**Comment by anthonyeleven** [comment]
Ack.  Several remaining unresolved suggestions, for either this PR or https://github.com/ceph/ceph/pull/51787

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 12 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:286]: <greg hirsch>If it is to be said, then let it be so.<\greg hirsch>

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 13 above)
Comment (zdover23) [reply doc/rados/configuration/bluestore-config-ref.rst:337]: See https://github.com/ceph/ceph/pull/51771#issuecomment-1564794725.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/51789 - Reef backport
https://github.com/ceph/ceph/pull/51790 - Quincy backport
https://github.com/ceph/ceph/pull/51791 - Pacific backport

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 12 above)
Comment (anthonyeleven) [reply doc/rados/configuration/bluestore-config-ref.rst:286]: <IronMaiden>So it shall be written; so it shall be done</IronMaiden>
<Mr. Wizard>Be what you is, and not what you is not.  Folks what do this is the happiest lot</Mr. Wizard>

File: doc/rados/configuration/bluestore-config-ref.rst (hunk 16 above)
Comment (anthonyeleven) [reply doc/rados/configuration/bluestore-config-ref.rst]: Excrement transpires.

**Pull request merged**